README.md
# git-proj

[//]: # (The README.html is generated from this file. Usually, pod is used for formatting.)

# Description

git-proj implements git sub-commands for managing large binary files.
Its setup and management is more straightforward than "git LFS" because it doesn't
*require* a git server.

[![test](https://github.com/TurtleEngr/gitproj/actions/workflows/test.yml/badge.svg)](https://github.com/TurtleEngr/gitproj/actions/workflows/test.yml)
[![package](https://github.com/TurtleEngr/gitproj/actions/workflows/package.yml/badge.svg)](https://github.com/TurtleEngr/gitproj/actions/workflows/package.yml)

## Why git-proj?

Have you ever had a developer commit their whole home directory into
your project's git repository (on purpose or accidentally)? I have
seen this. That includes their .ssh/ dir with their ssh keys. They
better have passwords on their private keys! Someone making this
mistake will likely have password-less keys. Yikes, what a security
mess--all their keys, and shared keys, will need to be changed on
all the systems they access.

Sure, you can "delete" the mistake, but if it were pushed to your main
git server, it would be in the repo forever unless you do the work
needed to rip it out. Yeah, some tools will do it, but they
are not trivial, and your repo should be "locked" for a few hours
until it is repaired. Maybe a faster solution: create a new repo from
a cleaned-up HEAD on all active branches, and delete the compromised
repo. Yeah, faster, but you've lost all your change history.

That is one scenario. This tool won't remove large binary
files from a git repo, but it can help with a more common scenario. Beginning
developers don't realize they shouldn't be committing object files,
exe files, log files, video files, or generated PDFs to the
repository. Your snappy repo will start to become huge and
*slow*. There are some good reasons for saving and tracking those kinds of files, but a git repo is not the place. So git-lfs was
created. You can skip this tool if you are OK with that solution and don't mind paying for
the storage. I wanted something simpler to
set up and with a more distributed approach. This tool is a start down
that path.

The main inspiration for this git-proj tool came from my desire to use
git to version video projects. But video files are HUGE, yet they
don't change much. Since most video files are rendered from video
editing tools, it is only important to version the video editor files to recreate a particular video file. Since rendering
takes time, you will want to save some rendered files. But there
is usually no need to save every rendered version.

Currently, this tool only supports git and large files on a local
system. The "remotes" can be set up on mounted drives, and they can be
used to clone the files to other systems.

See the
[issues](https://github.com/TurtleEngr/gitproj/issues?q=is%3Aopen+is%3Aissue+milestone%3A%222.0+Release%22),
milestone 2.0 tagged issues for future enhancements that will support
remotes across a network.

# git proj sub-commands

* [`git proj`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/git-proj.md) - general help.

* [`git proj init`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-init.md) - initialize a local git-proj repo.

* [`git proj-remote`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-remote.md) - initialize a new "remote repo" from the local one.

* [`git proj push`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-push.md) - push files in raw/ to the remote repo (and do a regular git push).

* [`git proj pull`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-pull.md) - pull raw/ files from the remote repo (and do a regular git pull).

* [`git proj status`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-status.md) - git status and status of binary files in raw/ compared to the remote's raw/.

* [`git proj clone`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-clone.md) - create a git workspace and raw/ workspace from a remote.

* [`git proj config`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-config.md) - Fix or change the git-proj configuration files.

* [`git proj add`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/subcommands/git-proj-add.md) - TBD. Add files and symlinks to the top raw/ directory.

[<img src="https://moria.whyayh.com/rel/released/software/own/git-proj/image/git-proj_User_DFD.png" width="600"/>](https://docs.google.com/presentation/d/15JqVxkBczpMKwtJI64FNU-2ixbXCh2ijwS3vfIBt4ko/edit#slide=id.p)

Click on the image above to see more details.

## User Docs

The user docs can be browsed online at:
[user-doc](https://github.com/TurtleEngr/gitproj/tree/main/doc/user-doc)

I recommend you start with:
[`git-proj`](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/git-proj.md)

Then the
[Create_a_git-proj_Repo](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/tutorial/create_a_git-proj_repo.md)
tutorial shows how to get started with the git-proj commands.

For a complete understanding of all the configuration variables and their
"levels," see
[gitproj Configuration Documentation](https://github.com/TurtleEngr/gitproj/blob/main/doc/user-doc/config.md).

# Installing

## System requirements:

* OS: Any Debian-based OS. For example, Ubuntu, Debian, Mint, MX. For
other OS, the "portable" installer, which ends with ".tar.gz", might
work. (Or see the [Quick
Start](https://github.com/TurtleEngr/gitproj/blob/develop/test/dev-doc/README.md#quick-start)
developer section for building a native package for your OS. This is a
harder path. I would recommend trying the portable installer first.)

* Packages: git (>=2.17), bash, coreutils (fmt, tr), gawk (awk),
git-flow, libpod-markdown-perl (pod2markdown), markdown,
openssh-client, openssh-server, perl (pod2html, pod2man, pod2usage),
rsync, sed, tidy. (If you use the apt-get package manager to install,
all of these packages will be installed when git-proj is installed.)

* The git sub-command directory /usr/lib/git-core exists.

## User requirements:

* You have used more than just `git clone`

* You have used the git CLI a lot (not just git in an IDE)

## Download Locations:

* The stable versions can be found at:
[https://moria.whyayh.com/rel/released/software/own/git-proj/deb](https://moria.whyayh.com/rel/released/software/own/git-proj/deb)
if prompted, use guest/guest for User/Password.

* The test packages (ones that have at least passed all tests) can be
found at:
[https://moria.whyayh.com/rel/development/software/own/git-proj/deb](https://moria.whyayh.com/rel/development/software/own/git-proj/deb)
if prompted, use guest/guest for User/Password.

[![test](https://github.com/TurtleEngr/gitproj/actions/workflows/test.yml/badge.svg)](https://github.com/TurtleEngr/gitproj/actions/workflows/test.yml)
[![package](https://github.com/TurtleEngr/gitproj/actions/workflows/package.yml/badge.svg)](https://github.com/TurtleEngr/gitproj/actions/workflows/package.yml)

Select the latest version (any OS). The ones with 'RC' or 'test' (and
timestamps) in the names are not "stable."

### Download example:

    tVer=0.7.1-1
    tPkg=git-proj-$tVer-mx-x86_64.deb
    tUrl=https://moria.whyayh.com/rel/released/software/own/git-proj/deb
    tOpt="--user=guest --password=guest"
    wget $tOpt $tUrl/$tPkg

### Install Example:

    sudo apt-get install ./$tPkg

# Getting more usage help

The user docs can be browsed online at:
[user-doc](https://github.com/TurtleEngr/gitproj/tree/main/doc/user-doc)

After installation, html and markdown help files and tutorials can
also be found at: /usr/share/doc/git-proj/user-doc/

Also, you can get help with these commands:

`man git-proj` - this will give you an overview of git-proj

`git proj -h` - output the overview AND usage help for ALL of the
git-proj sub-commands.

`git proj [CMD] -h` - output the help for "git proj CMD".
For example: "git proj init -h"

`git proj [CMD] -H usage` - this outputs just the short usage help for
"git proj CMD." For example: "git proj init -H usage." If a command
has required options, you can type the command for the usage
help.

# Source

[https://github.com/TurtleEngr/gitproj](https://github.com/TurtleEngr/gitproj)

# For Developers

The developer docs can be found at:
[test/dev-doc](https://github.com/TurtleEngr/gitproj/tree/develop/test/dev-doc)


CHANGES.md
# git-proj CHANGES

[//]: # (The CHANGES.html is generated from this file. Usually pod is used for formatting.)

----------

## tag-0.7.6

Sat Mar 26 08:00:36 PDT 2022

* Removed -C from rsync copy of raw/ files

* INT: Cleaned up formatting

* Added links to docs

* INT: Describe possible designs for remore raw methods. Issue #18

* INT: Grammerly checks run against all doc and usage text

* INT: Fixed up product deps. Fixed up docs a bit.

* Added Quick Start

* INT: Fixed download path for tar files.

* INT: Remove symlinks to tar files. With gpTestEnv and mTestEnv, specify the directory for the tar files.

* INT: Added markdown package

* Added index files to docs

* INT: Cleaned up the doc files so they are all generated consistently.

* Updated overview image

* Added UserDFD image and link to docs

----------

## tag-0.7.3

Thu 03 Mar 2022 11:23:08 AM PST

* Removed vars: bin and doc. Fixed symlinks, for moved files, to be relative. See issue #51

* Make sure git config var name are always lower case. Issue #2

* Updated the documentation with more examples. See  issue #30

* INT: Regenerated the test-env tar files, and repaired tests

* INT: Increased sleep time, so the "test Action" will work

* INT: Fixed tests that use upper-case in key names. Issue #2

* INT: Adding more tests for git-proj-config. Issue #15

* INT: Fixed pre-commit check for binary files.

----------

## tag-0.7.1

Wed 23 Feb 2022 10:49:07 PM PST

* Updated docs, fixed typos

* Implemented git-proj-config. It is mostly ready, but still rough in spots. Issue #15

* Documented the menus in git-proj-config. Issue #15

* INT: Added git-proj-config to git-proj help list. Issue #15

* INT: Wrote fComMenu. It is more general than fComSelect.

* INT: Added some tests for git-proj-config. More are needed. See Issue #31

* INT: Fixed pre-commit check for binary files. Grep changed.

* INT: Fixed up fComConfigCopy, so that the "to" file doesn't need to exist.

* INT: Updated stats.sh. It is still not very accurate.

* INT: Add update of doc/VERSION to be in "first" target for package/Makefile

* INT: documented the "epm" dependency for development. I.e. where to get it.

* INT: Regenerated the tar files, and repaired the tests.

* INT: Remove try/catch functions. They don't work well enough to use.

* INT: Fixed the dev-doc generation.

----------

## tag-0-5-7-1

Wed 09 Feb 2022 07:37:27 PM PST

* INT: Created fix-rel-links.sh and fixed LINK{} tags. Issue: #5

* Implement the pre-commit gitproj.hook.check-for-tabs. See issue #4

* INT: The code has been converted to use the fComYesNo function. Issue #9

* Get pre-commit from more than one place. See "config" document, and issue #16

* INT: Added fComMkPreCommit to init and clone. Implements issue #16

* INT: Updated docs. Reorganized build steps in package.

* INT: Moved scripts to test/utils. Uses clean-git-proj-repo.sh and clean-git-proj-repo-cron on release server.

* Push test packages to moria. See: [/rel/development/software/own/git-proj/deb/](https://moria.whyayh.com/rel/development/software/own/git-proj/deb/)

* INT: Document the release-server setup. Added rush documentation.

* Added "ff-only" to "git proj pull"

* INT: Implemented GitHub Actions: test, package. See Issue #27

    * INT: Get epm package that does not include mkver.pl. Issue #27
    * INT: Fixed released path. Issue #27
    * INT: Changed package.yml so it only runs if test succeeds. Issue #27
    * INT: Added sudo to apt-get. Issue #27
    * INT: Added epm and epm-helper requirements. Issue #27
    * INT: Simplified install-deps
    * INT: Added github actions badge

----------

## tag-0-5-5-2

Thu 03 Feb 2022 11:42:47 AM PST

* Updated docs

* Added 'why' section to README. Spell corrections.

* Updated most github links, in docs, to point to the "main" branch.

* INT: Renamed release VERSION file to REL-VERSION

* INT: Issue #23 fixed in version 0.5.5

* INT: Got Travis-CI working

* INT: Fixed tar file deps for tests

* INT: Skipped syslog tests on Travis-CI server

----------

## tag-0-5-4-1

Mon 31 Jan 2022 09:56:47 AM PST

* Updated docs

* Created allcreate_a_git-proj_repo tutorial

* Added raw/.remote.proj

* Changed default remote-min-space to 2g

* Renamed allow-tabs to check-for-tabs

* Changed: binary-file-size-limit to binary-file-size

* Added rm-trailing-sp, bash-fmt, and bash-lint to
      /usr/share/doc/git-proj/contrib/

* Create CODE_OF_CONDUCT.md

* Created issue template: Bug Report

* Created issue template: Feature Request

* Added "requires" to package. See README "System requirements"
    section.

* Changed syslog default to `false`

* INT: Spell checked docs

* INT: Cleaned up mount-dir issues

* INT: Fixed POD error

* INT: Passed pRemoteGitDir. It was not validating enough.

* INT: Changed coding TBD to TODO, so it does not conflict with the
  TBD in variables.

* INT: Created: test/dev-doc/coding-convention.pod

* INT: Created: test/dev-doc/outline.pod

* INT: Moved TODO.md to test/dev-doc/

* INT: Created directory design docs: test/dev-doc/enhancements/

* INT: Added archive target

* INT: Cleaned up output and prompt/response from 'git proj remote'

* INT: Renamed gpRemoteRawDir to gpRemoteRawOrigin

* INT: Renamed remote-raw-dir to remote-raw-origin

* INT: Refactored all of the gitproj config files. Mainly removed the
need for HOST config files.

* INT: Fixed up the handling of fLog, fError, gpVerbose, and common
  options

* INT: Created: Function Map in outline.pod

* INT: Removed unused var cCurDir

* INT: Added pull request template

* INT: Moved LICENSE file.

* INT: Patch VERSION in config files, for test setup..

* INT: Make sure first option is not a flag, in git-pro

----------

## tag-0-2-0-1

Tue Jan 11 12:03:36 2022 -0800

* Updated documentation

* Fixed git-proj so it calls sub-commands.

* Formatting changes and spell checked.

* Bold markdown does nothing in github, so use back-ticks to highlight

* Added CHANGES.md.

* Updated README.md.


test/stats.sh
#!/bin/bash

echo "For accurate counts 'make dist-clean' should be run before this."

cd ..

tProjStart=2021-10-23
tProjEnd=$(date +%F)
tProjDuration=$(dateutils.ddiff $tProjStart $tProjEnd)

tNumTests=$(grep '^test' test/test-*.sh | grep '()' | wc -l)
tNumAsserts=$(grep assert test/*.sh | wc -l)

# Includes all files, even most generated files.
# Excludes: binary files, package built files, and files I didn't write.
tTotalLines=$(cat $(find * -type f | grep -Ev 'package/dist|package/pkg|test/.ran-*|*.tgz|*.gif|shunit2*') | wc -l)

tLinesInTest=$(cat $(find test -type f) | wc -l)

tLinesInDoc=$(cat $(find doc -type f) | wc -l)
tWordsInDoc=$(cat $(find doc -type f) | wc -w)

tDoc=$(cat $(find * -prune -type f) | wc -l)
let tLinesInDoc+=tDoc
tDoc=$(cat $(find * -prune -type f) | wc -w)
let tWordsInDoc+=tDoc

tLinesInCode=$(cat $(find git-core -type f) | wc -l)
tDoc=$(
    awk '
        /=pod/,/=cut/ {
            print $0
        }
        /=internal-pod/,/=internal-cut/ {
            print $0
        }
    ' $(find git-core/* -type f) | wc -l
)
#let tLinesInDoc+=tDoc
let tLinesInCode-=tDoc

tDoc=$(
    awk '
        /=pod/,/=cut/ {
            print $0
        }
        /=internal-pod/,/=internal-cut/ {
            print $0
        }
    ' $(find test/* git-core/* -type f) | wc -w
)
let tWordsInDoc+=tDoc

tNumFun=$(grep '()' git-core/* | grep -Ev '#' | wc -l)
tNumCmds=$('ls' git-core/git-proj-* | wc -l)

let tLinesPerWeek=tTotalLines*7/tProjDuration

let tPagesInDoc=tWordsInDoc/250

# --------------------
cat <<EOF
Tests
        Number of tests:    $tNumTests
        Number of asserts:  $tNumAsserts
        Lines in test/:     $tLinesInTest
Doc
        Lines in doc:       $tLinesInDoc
        Word  in doc:       $tWordsInDoc
	Pages of doc:	    $tPagesInDoc
Code
        Number of git SubCmds: $tNumCmds
        Number of Functions:   $tNumFun
        Lines of code:         $tLinesInCode

Total for all:   $tTotalLines

Time: $tProjDuration days
Lines/Week: $tLinesPerWeek
EOF


test/test-com2.sh
#!/bin/bash

# ========================================
fUsage()
{
    fComUsage -s usage -f $cTestCurDir/test-com2.sh

    # This is the start of the testing internal documentation. See:
    # fGitProjComInternalDoc
    return

    cat <<\EOF >/dev/null
=pod

=for text ========================================

=for html <hr/>

=head1 test-com.inc

test-com2.inc - Common functions used in the test scripts

=head1 SYNOPSIS

    ./test.com2.sh [test,test,...]

=head1 DESCRIPTION

test-com2.sh uses test-env.tgz to create a test environment with a
HOME defined, so that existing config files in a real HOME directory
are not affected. (This was derived from from setUp in test-init.sh)

shunit2.1 is used to run the unit tests. If no test function names are
listed, then all of the test functions will be run.

=head1 RETURN VALUE

0 - if OK

=head1 ERRORS

=head1 EXAMPLES

=head1 ENVIRONMENT

=head1 FILES

=head1 SEE ALSO

shunit2.1

=head1 NOTES

=head1 CAVEATS

=head1 DIAGNOSTICS

=head1 BUGS

=head1 RESTRICTIONS

=head1 AUTHOR

=head1 HISTORY

=cut

EOF
}

# ========================================

# --------------------------------
setUp()
{
    unset cGetOrigin cGetTopDir cGitProjVersion cPID gErr

    unset gpAction gpAuto gpAutoMove gpBin \
        gpDoc gpFacility gpGitFlow gpHardLink gpLocalRawDir \
        gpLocalRawDirPat gpLocalRawSymLink gpLocalTopDir gpMaxSize \
        gpPath gpProjName gpSysLog gpVer gpVerbose

    local tTarIn=$gpTestEnv/test-env_ProjLocalDefined.tgz

    fTestSetupEnv
    fTestCreateEnv

    cd $HOME >/dev/null 2>&1
    if [ ! -r $tTarIn ]; then
        echo "Missing: $tTarIn [$LINENO]" 1>&2
	exit 1
    fi
    tar -xzf $tTarIn
    fTestPatchPath

    cp $gpDoc/config/gitconfig /tmp/gitconfig.sav

    gpVerbose=3
    gpUnitDebug=0
    return 0

    cat <<EOF >/dev/null
=internal-pod

=internal-head3 setUp

Before each test runs, this restores all of the script's initial
global variable settings,

=internal-cut
EOF
} # setUp

tearDown()
{
    # fTestRmEnv
    cp /tmp/gitconfig.sav $gpDoc/config/gitconfig
    gpUnitDebug=0
    if [ -n "$cHome" ]; then
        HOME=$cHome
    fi
    cd $gpTest >/dev/null 2>&1
    return 0
} # tearDown

# ========================================

fLookFor()
{
    local pVar=$1
    local pFile=$2

    if ! grep -q '\[com "test"\]' $pFile; then
        return 1
    fi
    timeout 2s grep -q "$pVar = defined" $pFile
    return $?
} # fLookFor

# --------------------------------
testComSetConfigMore()
{
    local tResult

    # Files:
    # Relative to $HOME ($cDatHome)
    #     1) -g ~/.gitconfig (include.path = .gitproj.config.global)
    # Relative to $HOME/$cDatProj1 ($cDatHome/$cDatProj1)
    #     2) -l .git/config
    #     X) -L .gitproj - with clone, updates --local
    #        after that it is updated from --local

    # ----------
    tResult=$(fComSetConfig -g -k com.test.gvar -v "defined" 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO" "fLookFor gvar $HOME/.gitconfig"

    # ----------
    cd $HOME/$cDatProj1
    tResult=$(fComSetConfig -l -k com.test.lvar -v "defined" 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO" "fLookFor lvar .git/config"

    # ----------
    cd $HOME
    tResult=$(fComSetConfig -l -k com.test.lvar -v "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "err: Error: You must be in a git workspace for this command"

    # ----------
    cd $HOME/$cDatProj1
    tResult=$(fComSetConfig -L -k com.test.Lvar -v "defined" 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO" "fLookFor lvar .gitproj"

    # ----------
    cd $HOME
    tResult=$(fComSetConfig -L -k com.test.Lvar -v "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "err: Error: You must be in a git workspace for this command"

    # ----------
    cd $HOME/$cDatProj1
    tResult=$(fComSetConfig -l -k com.test.lvar -v "defined" 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO" "fLookFor lvar $c .git/config"

    # ----------
    cd $HOME
    tResult=$(fComSetConfig -l -k com.test.Hvar -v "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "err: Error: You must be in a git workspace for this command"

    # ----------
    cd $HOME/$cDatProj1
    tResult=$(fComSetConfig -f $HOME/test-config -k com.test.fvar -v "defined" 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO" "fLookFor fvar $HOME/test-config"
    assertContains "$LINENO" "$tResult" "warning"
    assertContains "$LINENO" "$tResult" "created"

    # ----------
    cd $HOME/$cDatProj1
    mkdir $HOME/readonly
    chmod a-w $HOME/readonly
    tResult=$(fComSetConfig -f $HOME/readonly/test-fail -k com.test.fvar -v "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertTrue "$LINENO" "[ ! -f $HOME/readonly/test-fail ]"
    assertContains "$LINENO" "$tResult" "warning"
    assertContains "$LINENO" "$tResult" "created"
    assertContains "$LINENO" "$tResult" "Could not create"

    # ----------
    tResult=$(fComSetConfig -g -v "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "fComSetConfig: missing key"
    assertContains "$LINENO $tResult" "$tResult" "err: Internal: Error: fComSetConfig: missing key"
    assertContains "$LINENO $tResult" "$tResult" "Stack trace"

    # ----------
    tResult=$(fComSetConfig -g -k com.test.gvar "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "fComSetConfig: missing value"

    # ----------
    tResult=$(fComSetConfig -g -k com.test.gvar -v 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "Value required for option"

    # ----------
    tResult=$(fComSetConfig -W -g -k com.test.gvar -v "defined" 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "fComSetConfig:: Unknown option"

    # ----------
    #rm $HOME/$c ConfigGlobal
    tResult=$(fComGetConfig -e -g -k com.test.Xvar 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "crit: Error: Unexpected: com.test.xvar is not defined!"

    return 0
} # testComSetConfigMore

# --------------------------------
testComConfigCopy()
{
    local tResult
    local tSrc=$gpDoc/config/gitconfig
    local tDstDir=$HOME/project/test
    local tDst=$HOME/project/test/config.test
    local tIncPat
    local tExcPat

    # fComConfigCopy [-f] [-s pSrc] [-d pDst] [-i pIncPat] [-e pExcPat]

    tResult=$(fComConfigCopy 2>&1)
    assertFalse "$LINENO" "$?"
    assertContains "$LINENO $tResult" "$tResult" "is missing or not readable"

    tResult=$(fComConfigCopy -s foo 2>&1)
    assertFalse "$LINENO" "$?"
    assertContains "$LINENO $tResult" "$tResult" "foo is missing or not readable"

    tResult=$(fComConfigCopy -s $tSrc 2>&1)
    assertFalse "$LINENO" "$?"
    assertContains "$LINENO $tResult" "$tResult" "is missing or not writable"

    tResult=$(fComConfigCopy -s $tSrc -d $tDst 2>&1)
    assertFalse "$LINENO" "$?"
    assertContains "$LINENO $tResult" "$tResult" "directory does not exist"

    mkdir $tDstDir
    touch $tDst

    tResult=$(fComConfigCopy -s $tSrc -d $tDst 2>&1)
    assertTrue "$LINENO" "$?"
    assertTrue "$LINENO" "[ -f $tDst.bak ]"
    assertTrue "$LINENO $(diff -ibBwZ $tSrc $tDst)" "diff -qibBwZ $tSrc $tDst"


    git config -f $tDst gitproj.config.local-status local-status-val
    git config -f $tDst gitproj.config.local-host local-host-val
    git config -f $tDst gitproj.config.remote-status remote-status-val
    git config -f $tDst gitproj.config.proj-name proj-name-val
    tResult=$(fComConfigCopy -f -s $tSrc -d $tDst \
        -i 'gitproj\.config\.local-|gitproj.config.proj-name' \
	-e "gitproj\.config\.local-host" 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertFalse "$LINENO local-status" "grep -q local-status-val $tDst"
    assertFalse "$LINENO proj-name" "grep -q proj-name-val $tDst"
    assertTrue "$LINENO local-host" "grep -q local-host-val $tDst"
    assertTrue "$LINENO remote-status" "grep -q remote-status-val $tDst"

    return 0
} # testComConfigCopy

testComConfigUpdateLocal()
{
    local tResult
    local tValue
    local tGlobal=$gpDoc/config/gitconfig
    local tGitProj=$HOME/$cDatProj1/.gitproj
    local tLocal=$HOME/$cDatProj1/.git/config

    cd $HOME
    tResult=$(fComConfigUpdateLocal 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "fComConfigUpdateLocal"

    cd $HOME/$cDatProj1
    mv .gitproj .gitproj.sav
    tResult=$(fComConfigUpdateLocal 2>&1)
    assertFalse "$LINENO $tResult" "$?"
    assertContains "$LINENO $tResult" "$tResult" "fComConfigUpdateLocal"
    # Make sure there will be a time difference with next tests
    sleep 1

    cd $HOME/$cDatProj1
    # Restore
    mv .gitproj.sav .gitproj
    # Clean
    rm .gitproj.bak* 2>/dev/null
    # Change
    fComSetConfig -l -k "gitproj.config.verbose" -v 3
    # Check
    tValue=$(fComGetConfig -L -k "gitproj.config.verbose" -d null)
    assertEquals "$LINENO" "2" "$tValue"

    # Test
    tResult=$(fComConfigUpdateLocal 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO $tResult" "[ -f .gitproj.bak ]"
    assertContains "$LINENO $tResult" "$tResult" "1 file changed"
    tValue=$(fComGetConfig -L -k "gitproj.config.verbose")
    assertEquals "$LINENO" "3" "$tValue"
    sleep 1

    gpVerbose=3
    cd $HOME/$cDatProj1
    tResult=$(fComConfigUpdateLocal 2>&1)
    assertTrue "$LINENO $tResult" "$?"
    assertTrue "$LINENO $tResult $(echo ; stat  -c %y ./.git/config ./.gitproj )" "[ ./.git/config -ot ./.gitproj ]"
    ##assertNull "$LINENO $tResult $(echo ; stat  -c %y ./.git/config ./.gitproj )" "$tResult"
    ##assertNull "$LINENO $(diff ./.git/config ./.gitproj)" "diff -q ./.git/config ./.gitproj"
    ##assertContains "$LINENO $tResult $(echo ; stat  -c %y ./.git/config ./.gitproj )" "$tResult" "uncomment to show"
    return 0
} # testComConfigUpdateLocal

# --------------------------------
testComGetConfigMore()
{
    startSkipping
    fail "TODO"
    # untar a git env., test in and out of git dir
    return 0
} # testComGetConfigMore

# --------------------------------
testComUnsetConfigMore()
{
    startSkipping
    fail "TODO"
    return 0
} #testComUnsetConfigMore

# ========================================
# This should be the last defined function
fTestRun()
{
    if [ ! -x $gpTest/shunit2.1 ]; then
        echo "Error: Missing: $gpTest/shunit2.1"
        exit 1
    fi
    shift $#
    if [ -z "$gpTestList" ]; then
        # shellcheck disable=SC1091
        . $gpTest/shunit2.1
        exit $?
    fi

    # shellcheck disable=SC1091
    . $gpTest/shunit2.1 -- $gpTestList
    exit $?

    cat <<EOF >/dev/null
=internal-pod

=internal-head3 fTestRun

Run unit tests for the common functions.

=internal-cut
EOF
} # fTestRun

# ====================
# Main

export gpTest cTestCurDir gpTestList gpCmdName

gpCmdName=${BASH_SOURCE##*/}

# -------------------
# Set current directory location in PWD and cTestCurDir
if [ -z "$PWD" ]; then
    PWD=$(pwd)
fi
cTestCurDir=$PWD

# -------------------
# Define the location of this script
gpTest=${0%/*}
if [ "$gpTest" = "." ]; then
    gpTest=$PWD
fi
cd $gpTest
gpTest=$PWD
cd $cTestCurDir

# -----
# Optional input: a comma separated list of test function names
gpTestList="$*"

# -----
. $gpTest/test.inc

fTestRun $gpTestList


package/ver.sh
# Input DEF file for: mkver.pl.  All variables must have "export "
# at the beginning.  No spaces around the "=".  And all values
# enclosed with double quotes.  Variables may include other variables
# in their values.

# Set this to latest version of mkver.pl (earlier DEF files should
# still work with newer versions of mkver.pl)
export MkVer="2.2"

export ProdName="git-proj"
# One word [-a-z0-9]
# Required
# %provides ProdName

export ProdAlias="git-proj"
# One word [-a-z0-9]

export ProdVer="0.7.7"
# [0-9]*.[0-9]*{.[0-9]*}
# Requires 2 numbers, 3'rd number is optional
# First, Major changes:
#        Older projects should work, but they may need to be upgraded.
#        Newer projects will not work with older versions of git-proj
# Second, Minor changes, fixes, additions. Things probably won't break
#        for projects with the same Major version number.
# Third, Internal refactoring changes, changes in test messages, doc changes
# %version ProdVer

export ProdRC=""
# Release Candidate ver. Can be one or two numbers.
# If set and RELEASE=1
  # %release rc.ProdRC

export ProdBuild="1"
# [0-9.]*
# Required
# If RELEASE=1, and ProdRC=""
  # %release test.ProdBuild

# Generated: ProdBuildTime=YYYY.MM.DD.hh.mm
  # If RELEASE=0, or empty, or unset, then use
  # current time (UTC): %Y.%m.%d.%H.%M
    # %release ProdBuildTime

export ProdSummary="git-proj implements a git subcommand for managing large binary files."
# All on one line (< 80 char)
# %product ProdSummary

export ProdDesc="git-proj mostly solves the issue with large binary files causing git repo 'bloat'. Large files are not versioned, they are only copied. This is a much simpler implementation than git-lfs."
# All on one line
# %description ProdDesc

export ProdVendor="TurtleEngr"
# Required
# %vendor ProdVendor

export ProdPackager="$USER"
# Required
# %packager ProdPackager

export ProdSupport="turtle.engr\@gmail.com"
# Appended to %vendor

export ProdCopyright=""
# Current year if not defined
# %copyright ProdCopyright

export ProdDate=""
# 20[0-9][0-9]-[01][0-9]-[0123][0-9]
# Current date (UTC) if empty

export ProdLicense="dist/usr/share/doc/git-proj/LICENSE"
# Required
# %license ProdLicense

export ProdReadMe="dist/usr/share/doc/git-proj/README.md"
# Required
# %readme ProdReadMe

# Third Party (if any) If repackaging a product, define these:
export ProdTPVendor=""
# Appended to
export ProdTPVer=""
# Appended to
export ProdTPCopyright=""
# Appended to %copyright

export ProdRelServer="moria.whyayh.com"
export ProdRelRoot="/rel"
export ProdRelCategory="software/own/$ProdName/deb"
# Generated: ProdRelDir=$ProdRelRoot/released/$ProdRelCategory
# Generated: ProdDevDir=$ProdRelRoot/development/$ProdRelCategory

# Generated: ProdTag=tag-ProdVer-ProdBuild
# (All "." in ProdVer converted to "-")

# Generated: ProdOSDist=mx
# Generated: ProdOSVer=19
# Generated: ProdOS=mx19
# Generated: mx=1
# Generated: mx19=1
#       OSDist  OSVer
# linux
#       deb
#       ubuntu  16,18
#       mx      19,20
#       rhes
#       cent
#       fc
# cygwin
#       cygwin
# mswin32
#       win     xp
# solaris
#       sun
# darwin
#       mac

# Generated: ProdArch
# i386
# x86_64

# Output file control variables. (Unused types can be removed.)
# The *File vars can include dir. names
# The *Header and *Footer defaults are more complete than what is
# shown here.

export envFile="ver.env"
export envHeader=""
export envFooter="export CurLogName=$USER\n"

export epmFile="ver.epm"
export epmHeader="%include $ProdOSDist.require"
export epmFooter="%include epm.list"

export hFile="ver.h"
export hHeader=""
export hFooter=""

export javaPackage="DIR.DIR.DIR"
export javaInterface="ver"
export javaFile="ver.java"
export javaHeader=""
export javaFooter=""

export csNamespace="Supernode"
export csClass="ver"
export csFile="ver.cs"
export csHeader=""
export csFooter=""

export makFile="ver.mak"
export makHeader=""
export makFooter="CurLogName = $USER\n"

export plFile="ver.pl"
export plHeader=""
export plFooter=""

export xmlFile="ver.xml"
export xmlHeader=""
export xmlFooter=""
